File: macos/Classes/FilePickerPlugin.m
#import "FilePickerPlugin.h"
#import <Cocoa/Cocoa.h>

@implementation FilePickerPlugin
+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
  FlutterMethodChannel* channel = [FlutterMethodChannel
      methodChannelWithName:@"file_picker"
            binaryMessenger:[registrar messenger]];
  FilePickerPlugin* instance = [[FilePickerPlugin alloc] init];
  [registrar addMethodCallDelegate:instance channel:channel];
}

- (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
  if ([@"pickFile" isEqualToString:call.method]) {
    NSOpenPanel* panel = [NSOpenPanel openPanel];
    [panel setCanChooseFiles:YES];
    [panel setCanChooseDirectories:NO];
    [panel setAllowsMultipleSelection:NO];
    
    NSArray* allowedExtensions = @[@"json", @"characterbook", @"character", @"race", @"chax"];
    [panel setAllowedFileTypes:allowedExtensions];
    
    if ([panel runModal] == NSModalResponseOK) {
      NSString* path = [[[panel URLs] firstObject] path];
      result(path);
    } else {
      result(nil);
    }
  } else {
    result(FlutterMethodNotImplemented);
  }
}
@end

File: macos/Runner/Plugins/file_picker/FilePickerPlugin.swift
import FlutterMacOS
import AppKit

public class FilePickerPlugin: NSObject, FlutterPlugin {
    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(
            name: "file_picker",
            binaryMessenger: registrar.messenger
        )
        let instance = FilePickerPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "pickFile":
            handlePickFile(call: call, result: result)
        default:
            result(FlutterMethodNotImplemented)
        }
    }
    
    private func handlePickFile(call: FlutterMethodCall, result: @escaping FlutterResult) {
        DispatchQueue.main.async {
            let panel = NSOpenPanel()
            panel.canChooseFiles = true
            panel.canChooseDirectories = false
            panel.allowsMultipleSelection = false
            
            if let args = call.arguments as? [String: Any],
               let extensions = args["fileExtension"] as? String {
                let fileTypes = extensions
                    .replacingOccurrences(of: ".", with: "")
                    .components(separatedBy: ",")
                    .filter { !$0.isEmpty }
                panel.allowedFileTypes = fileTypes
            } else {
                panel.allowedFileTypes = ["json", "characterbook", "character", "race", "chax"]
            }
            
            let response = panel.runModal()
            if response == .OK, let url = panel.urls.first {
                result(url.path)
            } else {
                result(nil)
            }
        }
    }
}